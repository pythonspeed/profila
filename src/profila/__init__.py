def example_function():
    return 1 + 1
