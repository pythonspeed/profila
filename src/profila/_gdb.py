"""
Run GDB as a sub-process, gathering stack info.

By continuously interrupting a process, it's possible to get repeated stack
trace samples.  When run with debug info added, Numba gives informative stack
traces to gdb, at least, even if not to other tools, so we can use this info to
get Numba stack traces.

TODO: Currently only profiles the main thread.
"""

import asyncio
from asyncio.subprocess import Process
from collections.abc import AsyncIterable
from dataclasses import dataclass
import os
from shlex import quote
from time import time
from typing import Optional, cast
import sys

from pygdbmi.gdbmiparser import parse_response


@dataclass
class Frame:
    file: str
    line: int


async def _read(process: Process) -> Optional[dict[str, object]]:
    assert process.stdout is not None
    data_bytes = await process.stdout.readline()
    if process.returncode is not None:
        raise ProcessExited()
    data = data_bytes.decode("utf-8").rstrip()
    try:
        return cast(dict[str, object], parse_response(data))
    except Exception as e:
        print("ERROR PARSING GDB MESSAGE:", e, file=sys.stderr)
        return None


async def _sample(process: Process) -> AsyncIterable[Optional[list[Frame]]]:
    assert process.stdin is not None
    while True:
        start = time()
        process.stdin.write(b"-exec-interrupt\n")
        await _read_until_done(process)

        process.stdin.write(b"-stack-list-frames --no-frame-filters 0 10\n")
        message = await _read_until_done(process)
        if "stack" not in message["payload"]:  # type: ignore
            # Bad read of some sort:
            yield None
        else:
            yield [
                Frame(file=f["fullname"], line=int(f["line"]))
                for f in message["payload"]["stack"]  # type: ignore
                if ("fullname" in f) and ("line" in f)
            ]

        process.stdin.write(b"-exec-continue\n")
        await _read_until_done(process)
        elapsed = time() - start
        await asyncio.sleep(max(0.010 - elapsed, 0))
        # print(time() - start)


class ProcessExited(Exception):
    """The profiled process has exited."""


async def _read_until_done(process: Process) -> dict[str, object]:
    """
    Read until a command is done, return its result dictionary.
    """
    assert process.stdin is not None
    while True:
        result = await _read(process)
        if result is None:
            continue
        assert isinstance(result, dict)
        if result["type"] == "output":
            print(result["payload"])
        if result["type"] == "result":
            return result
        if result["type"] == "notify" and result["message"] == "thread-group-exited":
            await exit_subprocess(process)
            raise ProcessExited()


async def read_samples(process: Process) -> AsyncIterable[Optional[list[Frame]]]:
    """
    Return async iterable of samples read from the process.

    Call on result of ``run_subprocess()`` or ``attach_subprocess()``.
    """
    try:
        async for sample in _sample(process):
            yield sample
    except ProcessExited:
        await process.wait()
        return


async def run_subprocess(
    python_cli_args: list[str],
) -> Process:
    """
    Run Python in a subprocess.
    """
    env = os.environ.copy()
    # Make sure we get useful info from Numba
    env["NUMBA_DEBUGINFO"] = "1"
    # Get subprocess info in a timely manner:
    env["PYTHONUNBUFFERED"] = "1"

    process = await asyncio.create_subprocess_exec(
        "gdb",
        "--interpreter=mi3",
        stdout=asyncio.subprocess.PIPE,
        stdin=asyncio.subprocess.PIPE,
        env=env,
    )
    assert process.stdin is not None

    process.stdin.write(b"-gdb-set mi-async\n")
    await _read_until_done(process)
    process.stdin.write(f"-file-exec-file {quote(sys.executable)}\n".encode("utf-8"))
    await _read_until_done(process)
    process.stdin.write(
        b"-exec-arguments "
        + b" ".join(
            [os.fsencode(quote(s).replace("\n", "\\n")) for s in python_cli_args]
        )
        + b"\n"
    )
    await _read_until_done(process)
    process.stdin.write(b"-exec-run\n")
    await _read_until_done(process)

    return process


async def attach_subprocess(pid: str) -> Process:
    """
    Attach to an existing Python subprocess.
    """
    process = await asyncio.create_subprocess_exec(
        "gdb",
        "--interpreter=mi3",
        stdout=asyncio.subprocess.PIPE,
        stdin=asyncio.subprocess.PIPE,
    )
    assert process.stdin is not None

    process.stdin.write(b"-gdb-set mi-async\n")
    await _read_until_done(process)
    process.stdin.write(b"-target-attach %s\n" % pid.encode("ascii"))
    await _read_until_done(process)
    process.stdin.write(b"-exec-continue\n")
    await _read_until_done(process)

    return process


async def exit_subprocess(process: Process) -> None:
    """Exit GDB."""
    assert process.stdin is not None
    process.stdin.write(b"-gdb-exit\n")
    await process.wait()
