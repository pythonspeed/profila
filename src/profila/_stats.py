"""
Aggregate call stacks.
"""

from collections import Counter, defaultdict
from dataclasses import dataclass, field
from typing import Optional
from ._gdb import Frame


@dataclass(frozen=True)
class FinalStats:
    """
    Final result when all stats are gathered.
    """

    total_samples: int
    percent_bad_samples: float
    percent_other_samples: float
    # Map path to mapping of line number to percentage.
    numba_samples: dict[str, dict[int, float]]

    def total_percent(self) -> float:
        """
        Add up all percentages, should always be approximately 100%.
        """
        result = self.percent_bad_samples + self.percent_other_samples
        for line_counts in self.numba_samples.values():
            result += sum(line_counts.values())
        if result == 0.0:
            # Got absolutely nothing, which is also valid!
            result = 100.0
        return result


@dataclass
class Stats:
    # Map Python filenames to per-line counts. Should be Numba samples only.
    path_to_line_counts: defaultdict[str, Counter[int]] = field(
        default_factory=lambda: defaultdict(Counter)
    )
    # Samples we couldn't parse:
    bad_samples: int = 0
    # Samples that weren't Numba based:
    other_samples: int = 0

    def total_samples(self) -> int:
        """
        Total number of all samples.
        """
        result = self.bad_samples + self.other_samples
        for line_counts in self.path_to_line_counts.values():
            result += sum(line_counts.values())
        return result

    def add_sample(self, sample: Optional[list[Frame]]) -> None:
        """
        Add a sample.
        """
        if sample is None:
            self.bad_samples += 1
            return

        for frame in sample:
            if frame.file.endswith(".py"):
                self.path_to_line_counts[frame.file][frame.line] += 1
                return

        self.other_samples += 1

    def finalize(self) -> FinalStats:
        """
        Calculate final stats for human rendering.
        """
        total_samples = self.total_samples()

        def to_percent(count: int) -> float:
            if total_samples == 0:
                return 0.0
            return round((count / total_samples) * 100, 1)

        percent_bad_samples = to_percent(self.bad_samples)
        percent_other_samples = to_percent(self.other_samples)
        numba_samples = {}
        for filename, counts in self.path_to_line_counts.items():
            filename_counts: dict[int, float] = {}
            numba_samples[filename] = filename_counts
            for line_number, count in counts.items():
                filename_counts[line_number] = to_percent(count)

        final_stats = FinalStats(
            total_samples=total_samples,
            percent_bad_samples=percent_bad_samples,
            percent_other_samples=percent_other_samples,
            numba_samples=numba_samples,
        )
        assert -5.0 < final_stats.total_percent() - 100 < 5.0
        return final_stats
