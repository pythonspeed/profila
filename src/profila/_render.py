"""
Render ``FinalStats`` to human-readable text.
"""

from io import StringIO
from linecache import getline

from ._stats import FinalStats


def render_text(stats: FinalStats) -> str:
    """
    Render stats to text.
    """
    result = StringIO()
    result.write(
        f"**Total samples:** {stats.total_samples} "
        + f"({stats.percent_other_samples}% non-Numba samples, "
        + f"{stats.percent_bad_samples}% bad samples)\n"
    )

    for filename, line_percents in stats.numba_samples.items():
        min_line = min(line_percents)
        max_line = max(line_percents)

        result.write(f"\n{filename} (lines {min_line} to {max_line}):\n\n```\n")
        for line_number in range(min_line, max_line + 1):
            code = getline(filename, line_number).rstrip()
            percent = line_percents.get(line_number, 0)
            if percent == 0:
                usage = "      "
            else:
                usage = f"{percent:>5}%"
            result.write(f"{usage} | {code}\n")
        result.write("```\n")

    return result.getvalue()
