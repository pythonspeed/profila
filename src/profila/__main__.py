"""
Run Profila as a command-line tool.
"""

from argparse import ArgumentParser, REMAINDER, RawDescriptionHelpFormatter, Namespace
import asyncio
from asyncio.subprocess import Process
from dataclasses import asdict
import json
from shutil import which
import sys

from ._gdb import run_subprocess, read_samples, attach_subprocess, exit_subprocess
from ._stats import Stats
from ._render import render_text

PARSER = ArgumentParser(prog="profila", description="A profiler for Numba.")
SUBPARSERS = PARSER.add_subparsers()
ANNOTATE_PARSER = SUBPARSERS.add_parser(
    "annotate",
    help="Run a Python program and annotate the Numba source code.",
    formatter_class=RawDescriptionHelpFormatter,
    description="""To profile "python example.py":

    python -m profila annotate -- example.py

To profile "python -m yourpackage --arg=123":

    python -m profila annotate -m -- yourpackage --arg=123
""",
)
ANNOTATE_PARSER.add_argument(
    "rest",
    nargs=REMAINDER,
    help="The arguments you'd usually pass to the Python command-line.",
)
ANNOTATE_PARSER.set_defaults(command="annotate")
ATTACH_AUTOMATED_PARSER = SUBPARSERS.add_parser(
    "attach_automated",
    help="Attach to an existing process, for use by the Jupyter extension.",
)
ATTACH_AUTOMATED_PARSER.add_argument(
    "pid",
    action="store",
    help="The process PID.",
)
ATTACH_AUTOMATED_PARSER.set_defaults(command="attach_automated")


async def get_stats(process: Process) -> Stats:
    stats = Stats()

    count = 0
    async for sample in read_samples(process):
        count += 1
        stats.add_sample(sample)
    assert stats.total_samples() == count

    return stats


def annotate_command(args: Namespace) -> None:
    """
    Run the ``anotate`` command.
    """
    if args.rest and args.rest[0] == "--":
        del args.rest[0]

    if not which("gdb"):
        raise SystemExit(
            "gdb not found, make sure it is installed, e.g. "
            "run 'apt install gdb' on Ubuntu."
        )

    async def main() -> Stats:
        process = await run_subprocess(args.rest)
        return await get_stats(process)

    stats = asyncio.run(main())
    final_stats = stats.finalize()
    print(render_text(final_stats))


def attach_automated_command(args: Namespace) -> None:
    """
    Run the ``attach_automated`` command.

    The other side of this logic is in the ``_ipython.py`` module.
    """

    async def stop_on_stdin_close(process: Process) -> None:
        loop = asyncio.get_event_loop()
        # The Jupyter side will signal it's finished running code by closing
        # stdin.
        await loop.run_in_executor(None, sys.stdin.read)
        await exit_subprocess(process)

    async def main() -> Stats:
        process = await attach_subprocess(args.pid)
        asyncio.create_task(stop_on_stdin_close(process))
        # Tell the Jupyter side it can start running code:
        sys.stdout.write(json.dumps({"message": "attached"}) + "\n")
        sys.stdout.flush()
        return await get_stats(process)

    final_stats = asyncio.run(main()).finalize()
    # The source code is only available inside the Jupyter process (it's cells,
    # not files on the filesystem), so do the source code loading over there.
    sys.stdout.write(
        json.dumps({"message": "stats", "stats": asdict(final_stats)}) + "\n"
    )
    sys.stdout.flush()


def main() -> None:
    args = PARSER.parse_args()
    if args.command == "annotate":
        annotate_command(args)
    elif args.command == "attach_automated":
        attach_automated_command(args)
    else:
        raise NotImplementedError()


if __name__ == "__main__":
    main()
