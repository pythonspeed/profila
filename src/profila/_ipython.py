"""
IPython/Jupyter magics.
"""
import ctypes
import json
import sys
import os
from subprocess import Popen, PIPE
from time import time

from ._stats import FinalStats
from ._render import render_text

from IPython.core.magic import Magics, magics_class, cell_magic
from IPython.core.display import display, Markdown

libc = ctypes.CDLL("libc.so.6")
prctl = libc.prctl
prctl.argtypes = [ctypes.c_int]
prctl.restype = ctypes.c_int
# From linux/prctl.h:
PR_SET_PTRACER = ctypes.c_int(0x59616D61)


@magics_class
class ProfilaMagics(Magics):
    """
    IPython/Jupyter magics.
    """

    @cell_magic  # type: ignore[misc]
    def profila(self, line: str, cell: str) -> None:
        """Run the cell under a profiler."""
        del line

        # Allow this process' children to attach via ptrace(), so that gdb works:
        prctl(PR_SET_PTRACER, ctypes.c_long(os.getpid()))
        try:
            self._run_profila(cell)
        finally:
            # Switch back to normal ptrace() policy:
            prctl(PR_SET_PTRACER, ctypes.c_long(0))

    def _run_profila(self, cell: str) -> None:
        start = time()
        profiler = Popen(
            [
                sys.executable,
                "-m",
                "profila",
                "attach_automated",
                str(os.getpid()),
            ],
            stdin=PIPE,
            stdout=PIPE,
        )
        # Wait for it to be ready:
        assert profiler.stdin is not None
        assert profiler.stdout is not None
        assert json.loads(profiler.stdout.readline().rstrip())["message"] == "attached"

        # Run the code:
        assert self.shell is not None
        self.shell.run_cell(cell)

        # Tell the subprocess it can exit:
        profiler.stdin.close()

        # JSON can't have integer keys, so we need to convert strings (line
        # numbers) back to integers.
        message = json.loads(profiler.stdout.readline().rstrip())
        for line_mappings in message["stats"]["numba_samples"].values():
            for line, pct in list(line_mappings.items()):
                del line_mappings[line]
                line_mappings[int(line)] = pct

        assert message["message"] == "stats"
        final_stats = FinalStats(**message["stats"])

        elapsed = time() - start
        text = f"**Elapsed:** {elapsed:.3f} seconds\n\n" + render_text(final_stats)
        display(Markdown(text))  # type: ignore[no-untyped-call]
