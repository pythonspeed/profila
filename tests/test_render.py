"""
Tests for ``profila._render``.
"""

from syrupy.assertion import SnapshotAssertion

from profila._stats import FinalStats
from profila._render import render_text


def test_render_text(snapshot: SnapshotAssertion) -> None:
    """
    Minimal test for ``render_text()``.
    """
    final_stats = FinalStats(
        total_samples=1000,
        percent_bad_samples=9.9,
        percent_other_samples=15.1,
        numba_samples={"scripts_for_tests/simple.py": {12: 35.0, 15: 40.0}},
    )
    assert render_text(final_stats) == snapshot
