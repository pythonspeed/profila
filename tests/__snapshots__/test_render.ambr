# serializer version: 1
# name: test_render_text
  '''
  **Total samples:** 1000 (15.1% non-Numba samples, 9.9% bad samples)
  
  scripts_for_tests/simple.py (lines 12 to 15):
  
  ```
   35.0% |         result[i] = (7 + timeseries[i] / 9 + (timeseries[i] ** 2) / 7) / 5
         |     for i in range(len(result)):
         |         # This should be cheaper:
   40.0% |         result[i] -= 1
  ```
  
  '''
# ---
